resmem.c

```diff
@@ -50,37 +52,295 @@ static inline int get_info_infile(char *fname, char *res, void *out)
+		} else if (startswith("SwapTotal", buf) == 0) {
+			sscanf(buf, "%*s%zu", &swaptotal);
+		} else if (startswith("SwapFree", buf) == 0) {
+			sscanf(buf, "%*s%zu", &swapfree);
+		}
+	}
+	fclose(fp);
+
+	mmusage = cgmemread(cg, "memory.usage_in_bytes");
+	mmtot = cgmemlimit(cg, "memory.limit_in_bytes");
+	if (memtotal < mmtot) {
+		mmtot = memtotal;
+	}
+
+	((res_mem_infoall_t *)out)->memfree = mmtot - mmusage;
+	((res_mem_infoall_t *)out)->memavailable = mmtot - mmusage + cache;
+	((res_mem_infoall_t *)out)->memtotal = mmtot;
+	((res_mem_infoall_t *)out)->active = active_anon + active_file;
+	((res_mem_infoall_t *)out)->inactive = inactive_anon + inactive_file;
+
+	swusage = cgmemread(cg, "memory.memsw.usage_in_bytes");
+	swtot = cgmemlimit(cg, "memory.memsw.limit_in_bytes");
+
+	if (swtot > 0 && swtot < swaptotal)
+		swaptotal = swtot;
+
+	if (swtot > 0 && swusage > 0) {
+		swusage = swusage - mmusage;
+		swapfree = (swusage < swaptotal) ? swaptotal - swusage : 0;
+	}
+
+	((res_mem_infoall_t *)out)->swaptotal = swaptotal;
+	((res_mem_infoall_t *)out)->swapfree = swapfree;
+
+	return 0;
+}
+
 /* Read resource information corresponding to res_id */
 int getmeminfo(int res_id, void *out, void *hint, int pid, int flags)
 {
 	char buf[MEMBUF_128];
 	FILE *fp;
 	int err = 0;
+	size_t active_anon, active_file, inactive_anon, inactive_file, cache,
+		swaptotal, swapfree, swtot, swusage, mmusage, mmtot, memtotal;
+	int ret = 0;
 
-	switch (res_id) {
+	char *cg = get_cgroup(pid, MEMCGNAME);
+
+	if (cg) {
+		clean_init(cg);
+	}
 
+	switch (res_id) {
+		/* if process is part of a cgroup then return memory info
+		 *for that cgroup only.
+		 */
 	case RES_MEM_FREE:
-		return get_info_infile(MEMINFO_FILE, "MemFree:", out);
+		if (cg) {
+			mmusage = cgmemread(cg, "memory.usage_in_bytes");
+			if (get_info_infile(MEMINFO_FILE, "MemTotal:",
+				&memtotal) == -1)
+				return -1;
+			mmtot = cgmemlimit(cg, "memory.limit_in_bytes");
+			if (memtotal < mmtot) {
+				mmtot = memtotal;
+			}
+
+			*(size_t *)out = mmtot - mmusage;
+			return 0;
+		} else {
+			return get_info_infile(MEMINFO_FILE, "MemFree:", out);
+		}
 
 	case RES_MEM_AVAILABLE:
-		return get_info_infile(MEMINFO_FILE, "MemAvailable:", out);
+		if (cg) {
+			mmusage = cgmemread(cg, "memory.usage_in_bytes");
+			if (get_info_infile(MEMINFO_FILE, "MemTotal:",
+				&memtotal) == -1)
+				return -1;
+
+			mmtot = cgmemlimit(cg, "memory.limit_in_bytes");
+
+			if (memtotal < mmtot) {
+				mmtot = memtotal;
+			}
+			cache = cgmemstat(cg, "cache");
+
+			*(size_t *)out = mmtot - mmusage + cache;
+			return 0;
+		} else {
+			return get_info_infile(MEMINFO_FILE, "MemAvailable:",
+				out);
+		}
 
 	case RES_MEM_TOTAL:
-		return get_info_infile(MEMINFO_FILE, "MemTotal:", out);
+		ret = get_info_infile(MEMINFO_FILE, "MemTotal:", out);
+		if (cg) {
+			mmtot = cgmemlimit(cg, "memory.limit_in_bytes");
+			if (ret != -1 && *(size_t *)out > mmtot) {
+				*(size_t *)out = mmtot;
```

Comment:
Probably want to introduce proper pointer casting macros if you intend to do this more often. Since this is non-standard compliant how you do it. :) See LXC again:
https://github.com/lxc/lxc/blob/41cd8a8d8c3e5fd357a4b2fcdceb7b97d70586a4/src/lxc/macro.h#L381